dotcom-rendering/src/components/AdPortals.importable.tsx

```diff
@@ -0,0 +1,162 @@
+	a.y === b.y
+
+const getIsPremium = () => Promise.resolve(false);
+
+const calculateAdPositions = (adSlots: RefObject<HTMLDivElement>[]): BridgetAdSlot[] =>
+	adSlots
+		.map(slot => slot.current)
+		.filter(isNonNullable)
+		.map(slot => new BridgetAdSlot({
+			rect: calculateAdPosition(slot),
+			isSquare: false,
+		}))
+
+const adsHaveMoved = (oldPositions: BridgetAdSlot[], newPositions: BridgetAdSlot[]): boolean =>
+	newPositions.some((newPosition, idx) => {
+		const oldPositionRect = oldPositions[idx]?.rect;
+
+		if (oldPositionRect !== undefined) {
+			return !positionsEqual(newPosition.rect, oldPositionRect);
+		} else {
+			return false;
+		}
+	});
+
+export const AdPortals = () => {
+	// Server-rendered placeholder elements for ad slots to be inserted into.
+	const [adPlaceholders, setAdPlaceholders] = useState<Element[]>([]);
+	// References to client-side rendered ad slots.
+	const adSlots = useRef<RefObject<HTMLDivElement>[]>([]);
+	// Positions of client-side rendered ad slots.
+	const [adPositions, setAdPositions] = useState<BridgetAdSlot[]>([]);
+	// The height of the body element.
+	const [bodyHeight, setBodyHeight] = useState(0);
+
+	/**
+	 * Setup Ads
+	 * 
+	 * After first render, and only after first render, check if the user has
+	 * the premium tier. If not, set up ads:
+	 * - Look for ad placeholder elements in the page and store them
+	 * - Set up references to client-side rendered ad slots
+	 * - Store the height of the body element
+	 */
+	useEffect(() => {
+		void getIsPremium().then((isPremium) => {
+			if (!isPremium) {
+				const placeholders = Array.from(
+					document.getElementsByClassName(
+						'ad-portal-placeholder',
+					),
+				);
+
+				setAdPlaceholders(placeholders);
+				placeholders.forEach((_, i) => adSlots.current[i] = createRef());
+			}
+		});
+
+		setBodyHeight(document.body.clientHeight);
+	}, []);
+
+	/**
+	 * Insert Ads
+	 * 
+	 * Once the list of ad placeholders is available, the last render should
+	 * have populated the client-side ad slots. We can use references to those
+	 * ad slots to calculate their positions, and pass these to the native
+	 * layer so that it can insert ads.
+	 */
+	useEffect(() => {
+		if (adPlaceholders.length !== 0) {
+			const newAdPositions = calculateAdPositions(adSlots.current);
+			void getCommercialClient().insertAdverts(newAdPositions);
+			setAdPositions(newAdPositions);
+		}
+		
+	}, [adPlaceholders]);
+
+	/**
+	 * Update Ads
+	 * 
+	 * Ask the native layer to update the positions of the adverts whenever the
+	 * body element's height changes, because this typically means the ad slot
+	 * positions will have changed.
+	 *
+	 * A resize observer is used for this. Because it makes use of a callback
+	 * function, which closes over the `bodyHeight` and `adPositions` states,
+	 * we need to re-create it when either of those dependencies change.
+	 * 
+	 * TODO: debounce this to prevent calling the native layer too often?
+	 */
+	useEffect(() => {
```

Comment:
I think there are a couple of drawbacks with this useEffect:
- we create an observer and run the clean up function (to disconnect the previous observer) on every change to `bodyHeight` and `adPositions`
- the observer is tightly coupled to component state


I wonder if we can update as follows:

1. Create the observer once on component mount
1. The observer function will call `setBodyHeight` on every body resize
1. The cleanup function of the 'onmount' useEffect will be called on component 'unmount' so we call `disconnect()` there - once
1. Then we have the option of a second useEffect with a dependency array of just `bodyHeight` to detect changes which can derive `adPositions` and call bridget if necessary. Or if we assume that each re-render is effectively an update to body height then we make `adPositions` derived state from `bodyHeight` so it is calculated each time.

If we absolutely need to read component state in the observer we could store the observer on a ref and call disconnect separately in an 'unmount' useEffect. We used a similar approach on the YoutubeAtomPlayer:
- [listeners set on refs](https://github.com/guardian/csnx/blob/6a8b0988df38aee1b6c873cf319708c5ca04755b/libs/%40guardian/atoms-rendering/src/YoutubeAtomPlayer.tsx#L508-L511)
- [cleanup function on unmount](https://github.com/guardian/csnx/blob/6a8b0988df38aee1b6c873cf319708c5ca04755b/libs/%40guardian/atoms-rendering/src/YoutubeAtomPlayer.tsx#L566)